src/librustc_codegen_ssa/mir/mod.rs

```diff
@@ -26,7 +26,7 @@ use self::operand::{OperandRef, OperandValue};
 pub struct FunctionCx<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> {
     instance: Instance<'tcx>,
 
-    mir: &'a mir::Body<'tcx>,
+    mir: Option<&'a mut BodyCache<&'a mir::Body<'tcx>>>,
```

Comment:
This is an `Option` because of the juggling that was needed in https://github.com/rust-lang/rust/pull/64736/commits/eba7504ea6f8791ea525858e3cf078e24f4c3be8#diff-e682b38b565035a82f29560c3b79144bR19

Reply:
Does that deserve a comment to explain? The usages of unwrap are safe everywhere else because it's reinserted once that function ends. It's just taken out to avoid &mut and & references to the same `FunctionCx`

Reply:
Could this take ownership of a `BodyCache` instead, thus averting the whole borrowing problem?

Reply:
The issue is that the function that constructs the `FunctionCx` also requires references to the `&Body` inside of the `BodyCache`. I keep running into errors like the following:
```
error[E0502]: cannot borrow `mir` as mutable because it is also borrowed as immutable
   --> src/librustc_codegen_ssa/mir/mod.rs:238:58
    |
185 | pub fn codegen_mir<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
    |                    -- lifetime `'a` defined here
...
219 |     let (landing_pads, funclets) = create_funclets(&mir, &mut bx, &cleanup_kinds, &block_bxs);
    |                                    ----------------------------------------------------------
    |                                    |               |
    |                                    |               immutable borrow occurs here
    |                                    argument requires that `mir` is borrowed for `'a`
...
238 |     let memory_locals = analyze::non_ssa_locals(&mut fx, &mut mir);
    |                                                          ^^^^^^^^ mutable borrow occurs here
```

The above example is from an attempt to remove `BodyCache` entirely from `FunctionCx` and just pass it through all the methods. I'm struggling a little to figure out how to appease the checks here since I'll only ever use `&Body`, and never mutate the reference that's being used inside of `BodyCache`, but since `BodyCache`'s `Cache` has to be mutably borrowed, then the entire `BodyCache` is as well, and then I can't get a shared reference out for `&Body` =/.

Reply:
I figured out the lifetime issues and this is now back to being simple (almost).